orc8r/tools/magma-manipulator/magma_manipulator/main.py

```diff
@@ -0,0 +1,175 @@
+    cfg_puller_thread.start()
+
+
+def put_event_after_timeout(event):
+    LOG.debug('Wait {sec} seconds for event {event}'.format(
+        sec=event['timeout'], event=event['pod_name']))
+    if event['timeout'] > EVENT_MAX_TIMEOUT:
+        LOG.error('Can not handle event for pod {pod_name}. Timeout expired'
+                  .format(pod_name=event['pod_name']))
+        return
+    t = threading.Timer(event['timeout'], lambda: events_queue.put(event))
+    t.start()
+
+
+def main():
+    gws_manager = gateways.GatewaysManager()
+    start_periodic_tasks(gws_manager.get_gateways())
+
+    while True:
+        try:
+            if not events_queue.empty():
+                event = events_queue.get()
+                gw_pod_name = event['pod_name']
+                gw = gws_manager.get_gateway(gw_pod_name)
+
+                LOG.info('Handle event for {gw_pod_name}'.format(
+                    gw_pod_name=gw_pod_name))
+
+                if not k8s_tools.is_pod_ready(CONF.k8s.kubeconfig_path,
+                                              CONF.k8s.namespace,
+                                              gw_pod_name):
+                    event['timeout'] *= 2
+                    put_event_after_timeout(event)
+                    continue
+
+                if not utils.is_gw_reachable(gw.get_ip(gw_pod_name)):
+                    event['timeout'] *= 2
+                    put_event_after_timeout(event)
+                    continue
+
+                if not utils.is_cloud_init_done(
+                        gw.get_ip(gw_pod_name),
+                        CONF.gateways.username,
+                        CONF.gateways.rsa_private_key_path):
+                    event['timeout'] *= 2
+                    put_event_after_timeout(event)
+                    continue
+
+                if magma_api.is_gateway_in_network(CONF.orc8r_api_url,
+                                                   gw.network,
+                                                   gw.id,
+                                                   CONF.magma_certs_path):
+                    magma_api.delete_gateway(CONF.orc8r_api_url,
+                                             gw.network, gw.network_type,
+                                             gw.id, CONF.magma_certs_path)
+
+                # get gw hardware id and challenge key
+                gw_uuid, gw_key = gw.get_uuid_and_key()
+
+                magma_api.register_gateway(CONF.orc8r_api_url,
+                                           gw.network, gw.network_type,
+                                           gw.id, gw_uuid, gw_key,
+                                           gw.name, CONF.magma_certs_path)
+
+                magma_api.apply_gateway_config(CONF.orc8r_api_url,
+                                               gw.network, gw.network_type,
+                                               gw.id, gw.get_config(),
+                                               CONF.magma_certs_path)
```

Comment:
We should just combine the `magma_api.register_gateway` and `magma_api.apply_gateway_config` calls. Any data that can be specified as configuration can instead be provided when we register the gateway. This helps ensure that we don't get into a bad state where registration succeeds, but the config update fails.